test/Microsoft.AspNet.Razor.Test/Parser/CSharp/CSharpImplicitExpressionTest.cs

```diff
@@ -17,6 +18,105 @@ public override ParserBase CreateCodeParser()
+                    { "foo?(abc)?.bar?(def)", "foo?(abc)?.bar?(def)", AcceptedCharacters.NonWhiteSpace, noErrors },
+
+                    // Bracket data
+                    { "foo??[", "foo", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?[", "foo?[", AcceptedCharacters.Any, missingEndBracketError(5) },
+                    { "foo?[bar", "foo?[bar", AcceptedCharacters.Any, missingEndBracketError(5) },
+                    { "foo?[0]", "foo?[0]", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?[<p>", "foo?[", AcceptedCharacters.Any, missingEndBracketError(5) },
+                    { "foo?[bar.<p>", "foo?[bar.", AcceptedCharacters.Any, missingEndBracketError(5) },
+                    { "foo??[bar<p>", "foo", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?[-1]?", "foo?[-1]", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?[abc]?[def", "foo?[abc]?[def", AcceptedCharacters.Any, missingEndBracketError(11) },
+                    { "foo?[abc]?[2]", "foo?[abc]?[2]", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?[abc]?.bar?[def]", "foo?[abc]?.bar?[def]", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?[abc]?(bar)?.baz", "foo?[abc]?(bar)?.baz", AcceptedCharacters.NonWhiteSpace, noErrors },
+                    { "foo?[abc?(gef?[-1])]", "foo?[abc?(gef?[-1])]", AcceptedCharacters.NonWhiteSpace, noErrors },
+                };
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(NullPropagationData_ParenthesisBracket))]
+        public void ParseBlockMethodParsesNullPropagationImplicitExpression_ParenthesiBracket(
+            string implicitExpresison,
+            string expectedImplicitExpression,
+            AcceptedCharacters acceptedCharacters,
+            RazorError[] expectedErrors)
+        {
+            // Act & Assert
+            ImplicitExpressionTest(
+                implicitExpresison,
+                expectedImplicitExpression,
+                acceptedCharacters,
+                expectedErrors);
+        }
+
+        public static TheoryData NullPropagationData_Dot
+        {
+            get
+            {
+                // implicitExpression, expectedImplicitExpression
+                return new TheoryData<string, string>
+                {
+                    { "foo?", "foo" },
+                    { "foo??", "foo" },
+                    { "foo?!", "foo" },
+                    { "foo?.", "foo?." },
+                    { "foo??.", "foo" },
+                    { "foo?.(abc)", "foo?." },
+                    { "foo?.<p>", "foo?." },
+                    { "foo?.bar", "foo?.bar" },
+                    { "foo?.bar<p>", "foo?.bar" },
+                    { "foo??.bar<p>", "foo" },
+                    { "foo?.bar(false)?.<p>", "foo?.bar(false)?." },
+                    { "foo?.bar(false)?.baz", "foo?.bar(false)?.baz" },
```

Comment:
- mix some tests like `foo??bar` in here
- avoid "foo", "bar", et cetera when adding new tests